quesma/clickhouse/clickhouse.go

```diff
@@ -482,8 +496,38 @@ func generateNonSchemaFieldsString(attrsMap map[string][]interface{}) (string, e
 	return nonSchemaStr, nil
 }
 
+// This function implements heuristic for deciding if we should add new columns
+func (lm *LogManager) shouldAlterColumns(table *Table, attrsMap map[string][]interface{}) (bool, []int) {
+	attrKeys := getAttributesByArrayName(AttributesKeyColumn, attrsMap)
+	alterColumnIndexes := make([]int, 0)
+	if len(table.Cols) < maxColumns {
+		// We promote all non-schema fields to columns
+		// therefore we need to add all attrKeys indexes to alterColumnIndexes
+		for i := 0; i < len(attrKeys); i++ {
+			alterColumnIndexes = append(alterColumnIndexes, i)
+		}
+		return true, alterColumnIndexes
+	}
+	if lm.ingestFieldStatistics == nil {
+		lm.ingestFieldStatistics = make(IngestFieldStatistics)
+	}
+	const percent50 = 0.5
```

Comment:
Very strict, would almost never work. E.g. if after one day, new column always appear, it will be rejected.

In first iteration, we can simplify that will do it after X counts. Where X is 100 or 1000.